File: src/main/java/kitchenpos/menus/domain/menu/MenuName.java
package kitchenpos.menus.domain.menu;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import java.util.Objects;

@Embeddable
public class MenuName {
  @Column(name = "name", nullable = false)
  private String name;

  protected MenuName() {}

  protected MenuName(String name, MenuPurgomalumClient menuPurgomalumClient) {
    if (Objects.isNull(name) || name.isEmpty()) {
      throw new IllegalArgumentException("메뉴 이름을 입력해주세요.");
    }

    if (menuPurgomalumClient.containsProfanity(name)) {
      throw new IllegalArgumentException("메뉴 이름에 비속어가 들어갈 수 없습니다.");
    }

    this.name = name;
  }

  public String getName() {
    return name;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    MenuName menuName = (MenuName) o;
    return Objects.equals(name, menuName.name);
  }

  @Override
  public int hashCode() {
    return Objects.hash(name);
  }
}

File: src/main/kotlin/kitchenpos/tobe/menu/domain/vo/MenuName.kt
package kitchenpos.tobe.menu.domain.vo

import jakarta.persistence.Column
import jakarta.persistence.Embeddable
import kitchenpos.tobe.menu.domain.MenuPurgomalumClient

@Embeddable
class MenuName private constructor(
    @Column(name = "name", nullable = false)
    val name: String,
) {
    companion object {
        fun from(
            name: String,
            menuPurgomalumClient: MenuPurgomalumClient,
        ): MenuName {
            require(menuPurgomalumClient.containsProfanity(name).not()) {
                throw IllegalArgumentException("메뉴의 이름에 욕설이 포함되어 있습니다.")
            }
            return MenuName(name)
        }
    }
}
